### PR TITLE
feat: tests and bug fixes `AgentState.state`

### LINKED ISSUE #966: ValueError: 'system' not found in provided AgentState
People,

My first experiment with creating an agent (I am up to my fifth now so this error reporting is just for academic interest or in case it is useful) - otherwise using the default setup:

**Describe the bug**
See the output below

**Please describe your setup**

- [ ] MemGPT version
  - What is the output of `memgpt version`? (eg "0.2.4")
  0.3.0

- [ ] How did you install memgpt?
  - `pip install pymemgpt`

- [ ] Describe your setup
  - What's your OS
  Fedora Linux

  - How are you running `memgpt`?
  tmux Terminal

I can't remember what the chat was about when this happened but now whenever I try to chat to this agent, I get the following straight away:

```
Using existing agent phi_rho_01
Traceback (most recent call last):
  File "/usr/local/bin/memgpt", line 8, in <module>
    sys.exit(app())
             ^^^^^
  File "/usr/local/lib/python3.11/site-packages/typer/main.py", line 328, in __call__
    raise e
  File "/usr/local/lib/python3.11/site-packages/typer/main.py", line 311, in __call__
    return get_command(self)(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/typer/core.py", line 778, in main
    return _main(
           ^^^^^^
  File "/usr/local/lib/python3.11/site-packages/typer/core.py", line 216, in _main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/typer/main.py", line 683, in wrapper
    return callback(**use_params)  # type: ignore
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/memgpt/cli/cli.py", line 543, in run
    memgpt_agent = Agent(agent_state, interface=interface())
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/memgpt/agent.py", line 184, in __init__
    raise ValueError(f"'system' not found in provided AgentState")
ValueError: 'system' not found in provided AgentState
```

### PR DESCRIPTION
**Pull request by cpacker**
**Please describe the purpose of this pull request.**

Another attempt to patch #966 (closes #966)

- [x] add tests for read/writes on `AgentState.state`
  - added to `tests/test_metadata_store.py` (maybe should be put elsewhere?)
- [ ] in case of an empty `agent_state.state`, read from the preset instead
- [x] add tests for loading with empty `agent_state.state` (should no longer error)
- [x] remove `create_agent_from_preset`, instead modify `Agent.__init__` to take optional `preset` arg

**How to test**

- [x] reproduce linked issue
- [x] confirm that this PR patches the issue

**Have you tested this PR?**

### Repro issue (0.3.3)

Roll back to version `0.3.3`:
```sh
git checkout 508679e
poetry install
```

Wipe existing home dir:
```sh
mv ~/.memgpt ~/.memgpt_backup
```

Try to repro based on #1070:
```sh
memgpt quickstart
```
```
📖 MemGPT configuration file updated!
🧠 model        -> ehartford/dolphin-2.5-mixtral-8x7b
🖥️  endpoint     -> https://api.memgpt.ai
⚡ Run "memgpt run" to create an agent with the new config.
```

```sh
memgpt run
```
```
  File "/Users/loaner/dev/MemGPT-2/memgpt/cli/cli.py", line 620, in run
    memgpt_agent = presets.create_agent_from_preset(
  File "/Users/loaner/dev/MemGPT-2/memgpt/presets/presets.py", line 82, in create_agent_from_preset
    assert preset_name == preset.name, f"AgentState preset '{preset_name}' does not match preset name '{preset.name}'"
AttributeError: 'NoneType' object has no attribute 'name'
```

### Repro issue (`main`)

Roll back to version to main (pre-PR #1058):
```sh
git checkout main
poetry install
```

Wipe existing home dir:
```sh
mv ~/.memgpt ~/.memgpt_backup
```

Try to repro based on #1070:
```sh
memgpt quickstart
```
```
📖 MemGPT configuration file updated!
🧠 model        -> ehartford/dolphin-2.5-mixtral-8x7b
🖥️  endpoint     -> https://api.memgpt.ai
⚡ Run "memgpt run" to create an agent with the new config.
```

```sh
memgpt run
```
```
🧬 Creating new agent...
->  🤖 Using persona profile 'sam_pov'
->  🧑 Using human profile 'basic'
🎉 Created new agent 'BenevolentUnicorn' (id=8cf1de2c-6861-404b-abde-cc39f0269db2)

Hit enter to begin (will request first MemGPT message)
```

@sarahwooders bug reported in #1070 was fixed sometime past the `0.3.3` tag, but pre-this PR.

### ~~Confirm that PR patches the issue~~ Ensuring that there was no regression in this PR

```sh
memgpt quickstart
```
```
📖 MemGPT configuration file updated!
🧠 model        -> ehartford/dolphin-2.5-mixtral-8x7b
🖥️  endpoint     -> https://api.memgpt.ai
⚡ Run "memgpt run" to create an agent with the new config.
```

```sh
memgpt run
```
```
🧬 Creating new agent...
->  🤖 Using persona profile 'sam_pov'
->  🧑 Using human profile 'basic'
🎉 Created new agent 'LovingNachos' (id=3633e9a0-ad69-4dd0-b0d7-2e53f3b1c000)

Hit enter to begin (will request first MemGPT message)
```

### CODE REVIEW

File: memgpt/agent.py [hunk 1]
```diff
@@ -167,37 +168,81 @@ def initialize_message_sequence(
...
+                    "human": preset.human,
+                    "system": preset.system,
+                    "functions": preset.functions_schema,
+                    "messages": None,
+                },
+            )
+
+        # An agent can also be created directly from AgentState
+        elif agent_state is not None:
+            assert preset is None, "Can create an agent from a Preset or AgentState (but both were provided)"
```
Comment (cpacker) [inline memgpt/agent.py:212]: @sarahwooders regarding:

> in case of an empty agent_state.state, read from the preset instead

If we wanted to support this, we could comment out this assert (`assert preset is None`), and instead do something like:
```python
if preset is not None and agent_state.state is None:
    agent_state.state = {
                    "persona": preset.persona,
                    "human": preset.human,
                    "system": preset.system,
                    "functions": preset.functions_schema,
                    "messages": None,
    },
```

Comment (cpacker) [reply memgpt/agent.py:212]: However, I'm not sure if this is really necessary, and I think the current setup* may be fine.

*you can either pass an `AgentState` to the init of `Agent`, or you can pass a `preset` + a few other things (e.g. `created_by`), but you can't pass both at the same time.

**Review by sarahwooders — APPROVED** [review]
lgtm!
